### ISSUE TITLE
Subscriptions ready fires before Data has arrived (React)

### ISSUE THREAD

**Issue by patrickml** [kadirahq/flow-router#392]
So i am having an issue with a subscription being ready before the data has fully been served to the client. I am using ReactJS as a templating engine and I am passing the a `user` property to child templates. When i come to a setting page I need to subscribe to a specific piece of the users account and display that information in input fields to be edited.

However,

if the data has not quite hit the client the subscription is being set to ready before the template has the data to render. I'm not sure if its react slowing things down or what but here are a few example snippets that have been striped down.

**LayoutTemplate**
```jsx
PortalLayout = React.createClass({
  mixins : [ReactMeteorData],
  getMeteorData () {
    return {
      user : Meteor.user()
    }
  },
componentDidMount() {
    let instance = this;
    instance.tracker = Tracker.autorun(function () {
      FlowRouter.watchPathChange();
      if (instance.isMounted()) {
        let holder = instance.state;
        holder.current = FlowRouter.current();
        instance.setState(holder);
      }
    });
  },
 componentWillUnmount () {
   this.tracker.stop()
  },
  render() {

    if(!this.data.user) {
      return <div></div>
    }

    return (
      
        <div className="main-container">
          { React.cloneElement(this.props.content, { user: this.data.user }) }
        </div>

    )
  }
```
**Child Template**
```jsx

InsuranceSettings = React.createClass({
  componentDidMount () {
    let instance = this;

    instance._tracker = Tracker.autorun(function  () {
      let user = Meteor.user();
      let holder = instance.state;

      if(user && user.documents) {
        holder.exempt = user.documents.exempt;
      }

      holder.ready = FlowRouter.subsReady();
      instance.setState(holder);

    });

  },
  componentWillUnmount() {
    this._tracker.stop();
  },
  getInitialState () {
    return {
      exempt : 'no',
      ready : FlowRouter.subsReady()
    }
  },
  setExempt (event) {
    let holder = this.state;
    holder.exempt = $(event.target).val();
    this.setState(holder);
  },
  render () {
    let instance = this;

    if(!instance.state.ready || (instance.props.user && instance.props.user.documents && !instance.props.user.documents.glExp ))
      return (<div></div>);

    return (
      <div className="settings page block" id="account-settings">
        <form className="centered" ref="form">
          <h2>Insurance</h2>
          <label>General Liability</label>
          < FileUpload uploader="documents" name="gl" file={instance.props.user.documents && instance.props.user.documents.gl || null } />
          <label>General Liability Expiration</label>
          <input type="text" name="glExp" data-date defaultValue={instance.props.user.documents && instance.props.user.documents.glExp && instance.props.user.glExpDate() || '' } />

          <button type="button" className="pull-right" >Update Account</button>
        </form>
      </div>
    )
  }
});

```

Since React's `defaultValue` only sets the value on load i have to check if the values exist and if they don't ` if(!instance.state.ready || (instance.props.user && instance.props.user.documents && !instance.props.user.documents.glExp ))` then I have to render nothing until they do. But this doesnt always work because if `glExp` is sent before `gl` then `glExp` will have a value but `gl` won't.

If anyone has any ideas solutions or if this is the wrong place to ask this please let me know.

Thank you.

**Comment by patrickml** [comment]
So after someone bringing up Tracking FlowRouter in a Tracker auto run it got me thinking `getMeteorData` is already tracked. So i moved the ready into `getMeteorData` which oddly enough is updated before my other tracker but for some reason seemed to fix the issue. Im not 100% on this since it doesnt make much sense but hey

![Idontknow](https://pbs.twimg.com/media/A6PSWTFCUAI7N-5.jpg)

```jsx
InsuranceSettings = React.createClass({
  mixins : [ReactMeteorData],
  getMeteorData () {
    return {
      ready : FlowRouter.subsReady()
    }
  },
  componentDidMount () {
    let instance = this;

    instance.tracker = Tracker.autorun(function  () {
      //We are using flow router to track updates to the user
      FlowRouter.subsReady();
      let holder = instance.state;

      if(instance.props.user && instance.props.user.documents) {
        holder.exempt = instance.props.user.documents.exempt;
      }

      instance.setState(holder);
    });

  },
  componentWillUnmount() {
    this.tracker.stop();
  },
  saveDocument () {
    //logic to save
  },
  getInitialState () {
    return {
      exempt : 'no'
    }
  },
  setExempt (event) {
    let holder = this.state;
    holder.exempt = $(event.target).val();
    this.setState(holder);
  },
  render () {
    let instance = this;

    if(!instance.data.ready)
      return (<Loading />);

    return (
      <div className="settings page block" id="account-settings">
        <form className="centered" ref="form">
          <h2>Insurance</h2>
          <label>General Liability</label>
          < FileUpload uploader="documents" name="gl" file={instance.props.user.documents && instance.props.user.documents.gl || null } />
          <label>General Liability Expiration</label>
          <input type="text" name="glExp" data-date defaultValue={instance.props.user.documents && instance.props.user.documents.glExp && instance.props.user.glExpDate() || '' } />

          <label>Are you exempt from Workers Comp?</label>
          <div className="select" data-required>
            <select name="exempt" defaultValue={instance.props.user.documents && instance.props.user.documents.exempt || ''} onChange={instance.setExempt} >
              <option value="no">No</option>
              <option value="yes">Yes</option>
            </select>
          </div>

          <button type="button" className="pull-right" onClick={this.saveDocument}>Update Account</button>
        </form>
      </div>
    )
  }
});

```

**Comment by arunoda** [comment]
Okay. Here's the issue. 
getMeteorData() does more stuff and when manually setting state.  There could be some conflict. So, I advice you to put flow router code on the `getMeteorData` itself. Then it'll work without any issues.

```js
PortalLayout = React.createClass({
  mixins : [ReactMeteorData],
  getMeteorData () {
    const ready = FlowRouter.subsReady();
    if(!ready) {
      return {};
    }

    return {
      user : Meteor.user()
    }
  },
  render() {

    if(!this.data.user) {
      return <div></div>
    }

    return (

        <div className="main-container">
          { React.cloneElement(this.props.content, { user: this.data.user }) }
        </div>

    )
  }
}
```

**Comment by patrickml** [comment]
Though that may work it would cause react to re render the entire template whenever data is changed. My Portal Layout template has a lot more code than just the one div, i removed all of the code to isolate the issue for this thread. I am going to try this though

`{ this.data.ready && React.cloneElement(this.props.content, { user: this.data.user }) || '' }`

Since each of my subtemplates rending a loading indicator if the data isnt ready i should be able to move that check up to the main template now. And with both the FlowRouter Check and the Update on users being in the same place hopefully it will take care of the issue. Thank you for your time Arunoda!

**Comment by arunoda** [comment]
In React, don't worry about rendering. It won't change the DOM.

**Comment by patrickml** [comment]
Well unless it needs to remove everything

```jsx
if(!this.data.user) {
      return <div></div>
    }
```

will remove everything unless the user data exists causing the screen to flash every time the user data changes correct?

**Comment by tomRedox** [comment]
The way I have handled this is to split my components in two:  I move all the data handling to a wrapper component where I subscribe to and get the data.  I then pass that data down to the original  component (which is now devoid of any data handling code).  In that child component I implement shouldComponentUpdate() and check in there whether the incoming data (from the props) contains any data.  If it does I render, if not I cancel the render and wait for the second render that occurs when the data finishes loading in the wrapper.  

That has allowed me to keep my data access code in the component (rather than the router) and also means that I can use transition animations on the child component.

**Comment by andrewlorenz** [comment]
sounds great @tomRedox, a code snippet or a gist to show your pattern would be really great :)  thanks

**Comment by tomRedox** [comment]
Hi @andrewlorenz , I'm afraid I'm not working on Meteor or React at all at the moment, but the concept I described should be demonstrated in this example project I did at the time: https://github.com/tomRedox/simpleCRM .  I'd add the proviso that I haven't looked at that code in a long while, so things may well have moved on (no, let me rephrase that, it's JS, things will have **definitely** have moved on.  **A lot!**).  Hope that's some use.

**Comment by andrewlorenz** [comment]
I'm now using react-router plus react-meteor-data instead of flow-router - works perfectly.
